scripts/inject-deterministic.py

```diff
@@ -4,14 +4,129 @@
 import time
 from mitmproxy import ctx
 
+import hashlib
+
 
 class AddDeterministic():
 
+    def get_script_with_nonche(self, script, nonce=""):
+        """
+         Given a nonce, getScriptWithNonce returns the injected script text with the nonce.
+        If nonce is an empty string, getScriptWithNonce returns the script block
+        without attaching a nonce attribute.
+        Some responses may specify a nonce inside their Content-Security-Policy,
+        script-src directive.
+        The script injector needs to set the injected script's nonce attribute to
+        open execute permission for the injected script.
+        """
+
+        script_with_nouce = "<script"
+
+        if not nonce == "":
+            script_with_nouce += ' nonce="%s"' % nonce
+
+        script_with_nouce += ">"
+        script_with_nouce += script
+        script_with_nouce += "</script>"
+
+        return script_with_nouce
+
+    def get_nonce_from_headers(self, header):
+        """
+        get_nonce_from_headers returns the nonce token from a
+        Content-Security-Policy (CSP) header's script source directive, or an empty
+        string if the CSP header's script source does not contain a nonce.
+        For more background information on CSP and nonce, please refer to
+        https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/
+        Content-Security-Policy/script-src
+        https://developers.google.com/web/fundamentals/security/csp/
+        """
+
+        csp_script_directives = self.get_csp_script_directives(header)
+        if not csp_script_directives == "":
+            tokens = csp_script_directives.split(" ")
+            for token in tokens:
+                if token.startswith("'nonce-"):
+                    token = token.lstrip("'nonce-")
+                    token = token.rstrip("'")
+                    return token
+        return ""
```

Comment:
We can simplify this by adding methods for getting the CSP directives and script sources, such as:

```python
def get_csp_directives(self, headers):
    csp = headers.get("Content-Security-Policy", "")
    return [d.strip() for d in csp.split(";")]

def get_csp_script_sources(self, headers):
    for directive in self.get_csp_directives(headers):
        if directive.startswith("script-src "):
            return directive.split()[1:]

def get_nonce_from_headers(self, headers):
    for source in get_csp_script_sources(headers)
        if source.startswith("'nonce-")
            return source.partition("'nonce-")[-1][:-1]
```